pkg/virt-launcher/virtwrap/cli/event.go

```diff
@@ -0,0 +1,92 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package cli
+
+import (
+	"fmt"
+
+	"libvirt.org/libvirt-go"
+)
+
+type DomainEventDeviceRemovedCallbacker struct {
+	connect        *libvirt.Connect
+	domain         *libvirt.Domain
+	registrationID int
+	callback       libvirt.DomainEventDeviceRemovedCallback
+	eventChan      <-chan interface{}
+}
+
+func NewDomainEventDeviceRemovedCallbacker(
+	con Connection,
+	virDomain VirDomain,
+	callback libvirt.DomainEventDeviceRemovedCallback, eventChan <-chan interface{}) *DomainEventDeviceRemovedCallbacker {
+
+	libvirtConnect := extractLibvirtConnect(con)
+	libvirtDomain := extractLibvirtDomain(virDomain)
+	if libvirtConnect != nil && libvirtDomain != nil {
+		return &DomainEventDeviceRemovedCallbacker{
+			connect:   libvirtConnect,
+			domain:    libvirtDomain,
+			callback:  callback,
+			eventChan: eventChan,
+		}
+	}
+	return nil
+}
+
+func (c *DomainEventDeviceRemovedCallbacker) Register() error {
+	id, err := c.connect.DomainEventDeviceRemovedRegister(c.domain, c.callback)
```

Comment:
Why are you extracting the internal connection?
LibvirtConnection has a method for [DomainEventDeviceRemovedRegister](https://github.com/kubevirt/kubevirt/blob/b6b76766c43ea0bb5ba6a5a021c4cb7b87611957/pkg/virt-launcher/virtwrap/cli/libvirt.go#L162).

[Example](https://github.com/kubevirt/kubevirt/blob/0f30a9397f34fe91ac0d3570fe6ab6bd1a069215/pkg/virt-launcher/notify-client/client.go#L430-L429) usage.

Reply:
Why no adding `Deregister` method to `LibvirtConnection`?

Reply:
I think `LibvirtConnection` tries to add a different abstraction than the one we need here.
It has "register" methods that add the callback to a pool, from which point if we want to remove it from the pool it requires a refactoring to identify which callback to remove (it is a list now). But we do not need it in a pool from the first place, so doing this overhead work is unnecessary.

Maybe you see this from a different angle I am not.

Reply:
So add new `RegisterWithNoTracking` and `Deregister` methods to `LibvirtConnection`. I believe `LibvirtConnection` should keep the implementation details of the de/registration. Doing unsafe conversions and using inner connection feels wrong.

Reply:
I think the current model introduced by `LibvirtConnection` is wrong and adding to it is making something bad worse.
The suggested solution extracts the event handling out of that one-in-all structure, performs the extraction from the interface (which being an interface is by itself totally wrong) in the same package that implements both. I think that is better than the alternative.

It is not safe to add to a pool, which can be drained by some other method and then lookup thinks up which may or may not be there already.
Searching for the callback requires changing the structure from a list of functions, to a map[string]function. In Go you cannot compare function addresses.

Reply:
Can you please explain why the current model is wrong?
I see that in [DomainEventDeviceRemovedRegister](https://github.com/kubevirt/kubevirt/blob/b6b76766c43ea0bb5ba6a5a021c4cb7b87611957/pkg/virt-launcher/virtwrap/cli/libvirt.go#L162) the registration is surrounded we reconnection and a check whether the connection is lost.
Moreover, I say any interaction with libvirt is surrounded with this connections checks. Can you please explain why do you think it is generally wrong/not suitable to the case in this PR?
Can you ple

Regarding the list, you have the `registrationID`, so I find it pretty simple converting the list to a map.

Reply:
The current strategy reconnects to libvirt in a lazy way and not as a per-requirement of a flow.
The re-connection re-registers all callbacks as a side effect.
Moreover, it requires a check after each command (to libvirt) to see if it failed for multiple reasons (does not have to be a real connection problem). In case of an error (of the command), it returns the error back and it is up to the caller to decide what to do. I think in most cases it will just fail the flow.

I find this strategy irrelevant to the operation that is needed in this change. If the connection is down for any reason, I prefer the flow to fail and allow the reconcile to do its stuff.
I can also add, that the operation that comes immediately after is `dom.MigrateToURI3` which has no "wrapper", it directly calls libvirt `MigrateToURI3` with no reconnect and check. In fact, any operation on the `cli.VirtDomain` is not included into this "strategy".

Reply:
I have exposed two operations through `LibvirtConnection` and explained the reasoning in the commit messages.

I must admit that it feels uncomfortable to touch `LibvirtConnection` and its interface, but I guess nothing is perfect.